### ISSUE TITLE
Allow shadow and local capture to declare with super types and specify actual types with annotations

### ISSUE THREAD

**Issue by liach** [SpongePowered/Mixin#304]
Extension to #284.

```java
public class Outer {
  Inner inner;

  Inner getInner() {
    return inner;
  }

  void setInner(Inner i) {
    inner = i;
  }
  private static class Inner {}
}
```
Apparently, you cannot shadow any of these 2 methods, shadow the `inner` field, nor can you capture local variables correctly when one local var is of `Inner` type.

An annotation to specify actual types while the method descs are replaced with the actual type's supertypes/accessor mixin ducks would be invaluable, as it allows more mixin functionality without Access Transformers.

**Comment by Mumfrey** [comment]
Duplicate of #284, this is solved by Shadow Classes. I explain why I won't put the decorator in the annotation in my response to the other issue.

**Comment by liach** [comment]
tl;dr. Thanks.
